kimchi/src/circuits/polynomials/keccak/circuitgates.rs

```diff
@@ -0,0 +1,211 @@
+
+//~ ##### `KeccakXor` - XOR constraints for 32-bit words
+//~
+//~ * This circuit gate is used to constrain that `in1` xored with `in2` equals `out`.
+//~ * This gate operates on the `Curr` row and the `Next` row.
+//~
+//~ It uses three different types of constraints
+//~ * copy    - copy to another cell (32-bits)
+//~ * plookup - xor-table plookup (4-bits)
+//~
+//~ The 4-bit crumbs are assumed to be laid out with `0` being the least significant crumb.
+//~ Given values `in1`, `in2` and `out`, the layout looks like this:
+//~
+//~ | Column |          `Curr`  |          `Next`  |
+//~ | ------ | ---------------- | ---------------- |
+//~ |      0 | copy     `in1`   | copy     `out`   |
+//~ |      1 |                  | copy     `in2`   |
+//~ |      2 |                  |                  |
+//~ |      3 | plookup0 `in2_0` | plookup4 `in2_4` |
+//~ |      4 | plookup1 `in2_1` | plookup5 `in2_5` |
+//~ |      5 | plookup2 `in2_2` | plookup6 `in2_6` |
+//~ |      6 | plookup3 `in2_3` | plookup8 `in2_7` |
+//~ |      7 | plookup0 `in1_0` | plookup4 `in1_4` |
+//~ |      8 | plookup1 `in1_1` | plookup5 `in1_5` |
+//~ |      9 | plookup2 `in1_2` | plookup6 `in1_6` |
+//~ |     10 | plookup3 `in1_3` | plookup7 `in1_7` |
+//~ |     11 | plookup0 `out_0` | plookup4 `out_4` |
+//~ |     12 | plookup1 `out_1` | plookup5 `out_5` |
+//~ |     13 | plookup2 `out_2` | plookup6 `out_6` |
+//~ |     14 | plookup3 `out_3` | plookup7 `out_7` |
+//~
+
+#[derive(Default)]
+pub struct KeccakXor<F>(PhantomData<F>);
+
+impl<F> Argument<F> for KeccakXor<F>
+where
+    F: PrimeField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::KeccakXor);
+    const CONSTRAINTS: u32 = 3;
+
+    // Constraints for Xor
+    //   * Operates on Curr and Next rows
+    //   * Constrain the decomposition of `in1`, `in2` and `out`
+    //   * The actual XOR is performed thanks to the plookups.
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        let mut constraints = vec![];
+
+        let out_sum = four_bit(env, 14);
+        let in1_sum = four_bit(env, 10);
+        let in2_sum = four_bit(env, 6);
+
+        // Check first input is well formed
+        constraints.push(in1_sum - env.witness_curr(0));
+        // Check second input is well formed
+        constraints.push(in2_sum - env.witness_next(1));
+        // Check output input is well formed
+        constraints.push(out_sum - env.witness_next(0));
+
+        constraints
+    }
+}
+
+/// Computes the decomposition of a 32-bit word whose most significant 4-bit crumb
+/// is located in the `max` column of `witness_next`. The layout is the following:
+///
+/// |        | max - 3 | max - 2 | max - 1 |     max |
+/// | ------ | ------- | ------- | ------- | ------- |
+/// | `Curr` |  crumb0 |  crumb1 |  crumb2 |  crumb3 |
+/// | `Next` |  crumb4 |  crumb5 |  crumb6 |  crumb7 |
+///
+fn four_bit<F: PrimeField, T: ExprOps<F>>(env: &ArgumentEnv<F, T>, max: usize) -> T {
```

Comment:
This name doesn't convey a lot of meaning to the reader of the code.  Maybe `four_bit_decomoposition` or `four_bit_sum` would be better?